### PR TITLE
ForwardDiff handling rewrite

### PR DESCRIPTION
**Pull request by chriselrod**
Also add `params` function for returning params.

Fixes https://github.com/PumasAI/DeepPumas.jl/issues/165#issuecomment-1106638075
Addresses https://github.com/PumasAI/SimpleChains.jl/issues/24

### CODE REVIEW

File: src/forwarddiff_matmul.jl [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+end
+function collapse_dims12(A::PtrArray{S,(true, true, true),T,3,1,0,(1, 2, 3)}) where {S,T}
+  M, N, P = size(A)
+  sp = stridedpointer(A)
+  o1, o2, o3 = offsets(sp)
+  x1, _, x3 = strides(sp)
+  si = ArrayInterface.StrideIndex{3,(1, 2, 3),1}((x1, StaticInt(0), x3), (o1, o2, o3))
+  spnew = stridedpointer(pointer(sp), si)
+  PtrArray(spnew, (M * N, StaticInt(1), P), Val((true, true, true)))
+end
```
Comment (korsbo) [inline src/forwarddiff_matmul.jl:109]: Maybe add a test that hits these functions?

File: src/forwarddiff_matmul.jl [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+    r = reinterpret_dual
+    DCN = DC
+    DAN = DA
+  else
+    r = reinterpret_reshape_dual
+    dimC::Int = Int(length(DC))
+    new_dim = ((Int(length(DA))::Int + Int(length(DB))::Int - dimC) >>> 1) + dimC
+    DCN = (new_dim, DC...)
+    DAN = (new_dim, DA...)
+  end
```
Comment (korsbo) [inline src/forwarddiff_matmul.jl:293]: Add test?

File: src/dense.jl [hunk 3]
```diff
@@ -71,6 +71,17 @@ function getparams(td::TurboDense{true}, p::Ptr{T}, inputdim::Integer) where {T}
   W = PtrArray(reinterpret(Ptr{T}, p), (outputdim, inputdimp1))
   W, p + (outputdim * inputdimp1) * sizeof(T)
 end
+# to support `params`
+function _getparams(layer::TurboDense{false}, p, inputdim::Tuple)
+  _getparams(layer, p, last(inputdim))
+end
```
Comment (korsbo) [inline src/dense.jl:77]: Add a test that covers getting parameters from networks that mix layers with and without bias?

File: src/utils.jl [hunk 4]
```diff
@@ -219,3 +216,33 @@ function alloc_threaded_grad(
...
+getparams(_, p, inputdim) = nothing, p
+_getparams(::Nothing, p, inputdim::Tuple) = nothing, p, outputdim
+function _getparams(layer, p, inputdim::Tuple)
+  A, p = getparams(layer, p, inputdim)
+  _, outputdim = layer_output_size(Val{Float32}(), layer, inputdim)
+  A, p, outputdim
+end
+
+
+function params(sc::SimpleChain, p::AbstractVector, inputdim = nothing)
```
Comment (korsbo) [inline src/utils.jl:230]: Add docstring? Seems like a function that the user might want to know about.

File: src/forwarddiff_matmul.jl [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+  Bref = Expr(:ref, :B)
+  for d in DB
+    i = findfirst(==(d), DC)
+    if i === nothing
+      r = findfirst(==(d), contract_dims)
+      @assert r !== nothing "contract dims must occur in both `A` and `B`"
+      dimsym = Rinds[r]
+    else
+      @assert d ∉ DA "non contracting elementwise mul not currently supported"
```
Comment (korsbo) [inline src/forwarddiff_matmul.jl:179]:
```suggestion
      d ∉ DA && throw(ArgumentError("non contracting elementwise mul not currently supported"))
```

**Comment by korsbo** [comment]
What's the status of this PR @chriselrod?

File: src/forwarddiff_matmul.jl (hunk 5 above)
Comment (chriselrod) [reply src/forwarddiff_matmul.jl:179]: Should be `||`, or `\in`, as we want `d` not to be in `DA`.

File: src/dense.jl (hunk 3 above)
Comment (chriselrod) [reply src/dense.jl]: Always a good idea to actually test things of course...

But, out of curiosity, did you try it and get an error?

I had to fix a bug for this to work.

File: src/utils.jl (hunk 4 above)
Comment (chriselrod) [reply src/utils.jl:234]: Added

File: src/forwarddiff_matmul.jl (hunk 2 above)
Comment (chriselrod) [reply src/forwarddiff_matmul.jl:256]: I think this should be hit now.

File: src/forwarddiff_matmul.jl (hunk 1 above)
Comment (chriselrod) [reply src/forwarddiff_matmul.jl]: These methods are no longer used as of this PR, so I've removed them.

**Comment by chriselrod** [comment]
> What's the status of this PR @chriselrod?

This should be good to go.
I've also finally diagnosed and fixed the cause of the sporadic test failures on CI.

File: src/dense.jl (hunk 3 above)
Comment (korsbo) [reply src/dense.jl]: Hehe, no I had not tested it. It just seemed like something that's easy to get wrong or break so I wanted tests.

File: src/forwarddiff_matmul.jl (hunk 5 above)
Comment (korsbo) [reply src/forwarddiff_matmul.jl:142]: Meh, I'd prefer `throw` over all the `@assert`s in here but it's too minor a point to warrant more back-and-forth. Good enough already!

**Review by korsbo — APPROVED** [review]
Looks good to me!

**Pull request merged**